pkg/gitreceive/build.go

```diff
@@ -0,0 +1,295 @@
+package gitreceive
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"code.google.com/p/go-uuid/uuid"
+	"github.com/deis/builder/pkg"
+	"github.com/deis/builder/pkg/log"
+	"gopkg.in/yaml.v2"
+)
+
+const (
+	shortShaIdx = 8
+)
+
+type errGitShaTooShort struct {
+	sha string
+}
+
+func (e errGitShaTooShort) Error() string {
+	return fmt.Sprintf("git sha %s was too short", e.sha)
+}
+
+// repoCmd returns exec.Command(first, others...) with its current working directory repoDir
+func repoCmd(repoDir, first string, others ...string) *exec.Cmd {
+	cmd := exec.Command(first, others...)
+	cmd.Dir = repoDir
+	return cmd
+}
+
+// mcCmd returns a command to execute the 'mc' binary, so that it reads config from configDir.
+// the command outputs its stderr to os.Stderr
+func mcCmd(configDir string, args ...string) *exec.Cmd {
+	cmd := exec.Command("mc", "-C", configDir, "--quiet")
+	cmd.Args = append(cmd.Args, args...)
+	cmd.Stderr = os.Stderr
+	return cmd
+}
+
+func kGetCmd(podNS, podName string) *exec.Cmd {
+	return exec.Command("kubectl", fmt.Sprintf("--namespace=%s", podNS), "get", "pods", "-o", "yaml", podName)
+}
+
+// run prints the command it will execute to the debug log, then runs it and returns the result of run
+func run(cmd *exec.Cmd) error {
+	cmdStr := strings.Join(cmd.Args, " ")
+	if cmd.Dir != "" {
+		log.Debug("running [%s] in directory %s", cmdStr, cmd.Dir)
+	} else {
+		log.Debug("running [%s]", cmdStr)
+	}
+	return cmd.Run()
+}
+
+func build(conf *Config, builderKey, gitSha string) error {
+	storage, err := getStorageConfig()
+	if err != nil {
+		return err
+	}
+	creds, err := getStorageCreds()
+	if err == errMissingKey || err == errMissingSecret {
+		return err
+	}
+
+	repo := conf.Repository
+	if len(gitSha) <= shortShaIdx {
+		return errGitShaTooShort{sha: gitSha}
+	}
+	shortSha := gitSha[0:8]
+	appName := conf.App()
+
+	repoDir := filepath.Join(conf.GitHome, repo)
+	buildDir := filepath.Join(repoDir, "build")
+
+	slugName := fmt.Sprintf("%s:git-%s", appName, shortSha)
+	imageName := strings.Replace(slugName, ":", "-", -1)
+	if err := os.MkdirAll(buildDir, os.ModeDir); err != nil {
+		return fmt.Errorf("making the build directory %s (%s)", buildDir, err)
+	}
+	tmpDir := os.TempDir()
+
+	tarURL := fmt.Sprintf("%s://%s:%s/git/home/%s/tar", storage.schema(), storage.host(), storage.port(), slugName)
+
+	// this is where workflow tells slugrunner to download the slug from, so we have to tell slugbuilder to upload it to here
+	pushURL := fmt.Sprintf("%s://%s:%s/git/home/%s/push", storage.schema(), storage.host(), storage.port(), fmt.Sprintf("%s:git-%s", appName, gitSha))
+
```

Comment:
@smothiki actually, I think it would be nicer if the controller API `/v2/hooks/build` could take the URL to assign to slugrunner. It would reduce the coupling between builder and the controller.